langchain4j-core/src/main/java/dev/langchain4j/rag/content/Content.java

```diff
@@ -23,19 +24,27 @@
 public class Content {
 
     private final TextSegment textSegment;
+    private final Map<ContentMetadata, Object> metadata;
```

Comment:
Warning: this is just a suggestion and it's very likely that I'm overthinking this.

I'm noticing that in a few places like this we have a Map of heterogeneous values, where each value must be of a type determined by its key, but at the moment we don't have a way to enforce the right type at compile time. E.g.: we cannot enforce that the key `ContentMetadata.SCORE` will have a `Double` as value.

A (probably overeengineered) solution)would be to use the Map implementation that I sketched here https://gist.github.com/mariofusco/c9071b51d18b5577d22f3e25cc9e003c using as enumeration a `sealed interface` and a set of singletons implementing it. 

@langchain4j Maybe something similar could be also (re)used for the builders refactor that we were discussing separately?

Reply:
I like having better type safety for sure, but AFAIU this approach will require users to introduce new types (that implement `ContentMetadata`) if they will want to add custom fields to their custom `ContentRetriever`s? We need to weight all pros and cons and also analyze what other places might benefit from this approach (to have a unified approach throughout the library. But this requires more time, so I propoge to go with this PR for now and rethink this closer to 1.0?

Reply:
I missed that currently the key is an enum, so your approach will be better for customization for sure